### PR TITLE
[Streamtape] Add a new extractor for streamtape

### PR DESCRIPTION
**Pull request by brpy**
## Please follow the guide below

- You will be asked some questions, please read them **carefully** and answer honestly
- Put an `x` into all the boxes [ ] relevant to your *pull request* (like that [x])
- Use *Preview* tab to see how your *pull request* will actually look like

---

### Before submitting a *pull request* make sure you have:
- [x] At least skimmed through [adding new extractor tutorial](https://github.com/ytdl-org/youtube-dl#adding-support-for-a-new-site) and [youtube-dl coding conventions](https://github.com/ytdl-org/youtube-dl#youtube-dl-coding-conventions) sections
- [x] [Searched](https://github.com/yt-dlp/yt-dlp/search?q=is%3Apr&type=Issues) the bugtracker for similar pull requests
- [x] Checked the code with [flake8](https://pypi.python.org/pypi/flake8)

### In order to be accepted and merged into youtube-dl each piece of code must be in public domain or released under [Unlicense](http://unlicense.org/). Check one of the following options:
- [ ] I am the original author of this code and I am willing to release it under [Unlicense](http://unlicense.org/)
- [x] I am not the original author of this code but it is in public domain or released under [Unlicense](http://unlicense.org/) (provide reliable evidence)

### What is the purpose of your *pull request*?
- [ ] Bug fix
- [ ] Improvement
- [x] New extractor
- [ ] New feature

---

### Description of your *pull request* and other information

Code taken from https://github.com/schnusch/youtube-dl/commit/041c4c40f741974e60754b1e777de8e8248a3a3c

Author : https://github.com/schnusch

### CODE REVIEW

File: yt_dlp/extractor/streamtape.py [hunk 1]
```diff
@@ -0,0 +1,54 @@
+from __future__ import unicode_literals
+
+from .common import InfoExtractor
+from ..utils import js_to_json, urljoin
+
+# strings are obfuscated by concatenating substrings
+split_string_part = r'(?:%s|%s)' % (r'"(?:[^"\\]|\\.)*"',
+                                    r"'(?:[^'\\]|\\.)*'")
```
Comment (pukkandan) [inline yt_dlp/extractor/streamtape.py:8]: cant backreferences be used to simply this?

File: yt_dlp/extractor/streamtape.py [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+split_string = r'(?:' + split_string_part + r'(?:\s*\+\s*' + split_string_part + r')*)'
+videolink = r"(?:'\+')?".join('videolink')
+videolink = r"document\.getElementById\('" + videolink + r"'\)\.innerHTML\s*=\s*(?P<data>" + split_string + r")"
```
Comment (pukkandan) [inline yt_dlp/extractor/streamtape.py:11]: Use more descriptive variable names

File: yt_dlp/extractor/streamtape.py [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+
+
+class StreamtapeIE(InfoExtractor):
+    _VALID_URL = r'https?://(?:www\.)?streamtape.com/[ev]/(?P<id>[^/?#]+)'
+    _TESTS = [{
+        'url': 'https://streamtape.com/v/AJD9gAVwMOcXrQ8/Big_Buck_Bunny_Trailer_400p.ogg',
```
Comment (pukkandan) [inline yt_dlp/extractor/streamtape.py:17]: This URL no longer exists. Write a new test

File: yt_dlp/extractor/streamtape.py [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+
+    def _real_extract(self, url):
+        video_id = self._match_id(url)
+
+        webpage = self._download_webpage(url, video_id)
+
+        video = self._html_search_regex(videolink, webpage, 'video', group='data')
+        video = video.split('+')
+        video = [self._parse_json(v, video_id, js_to_json) for v in video]
+        video = urljoin(url, ''.join(video))
```
Comment (pukkandan) [inline yt_dlp/extractor/streamtape.py:36]: Don't reuse the same variable name again and again. Or even better, simplify this into 1 or 2 lines

File: yt_dlp/extractor/streamtape.py [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+
+        try:
+            poster = self._html_search_regex(r' id="mainvideo"[^>]* poster="(?P<data>.*?)"',
+                                             webpage, 'poster', group='data')
+            poster = urljoin(url, poster)
+        except ValueError:
+            poster = None
```
Comment (pukkandan) [inline yt_dlp/extractor/streamtape.py:43]: Don't use `try...except`
`_html_search_regex` has a `fatal` parameter

File: yt_dlp/extractor/streamtape.py (hunk 1 above)
Comment (brpy) [reply yt_dlp/extractor/streamtape.py:8]: I am not good with regex. Could you may be suggest changes?

**Comment by llacb47** [comment]
Streamtape is used for pirated content fyi.

File: yt_dlp/extractor/streamtape.py (hunk 3 above)
Comment (brpy) [reply yt_dlp/extractor/streamtape.py]: Wrote new test

**Comment by brpy** [comment]
> Streamtape is used for pirated content fyi.

Hi, I saw this pr not being merged in youtube-dl stating the same reason. 

I thought yt-dlp adds support for sites more quicker. Is adding support for *such* sites a *no go* in yt-dlp?

I use this site for legitimate videos and I wanted to use youtube-dl with it.

**Comment by pukkandan** [comment]
> Streamtape is used for pirated content fyi.

Thanks for the headsup @llacb47 

> Is adding support for such sites a no go in yt-dlp?

Yes @brpy, Piracy sites are a no-go.
Trust me, I want to merge all PRs that comes this way, but I don't want to risk this project (or myself) getting into any legal trouble. So if the site has lots of pirated content, it won't be officially supported.

> I use this site for legitimate videos and I wanted to use youtube-dl with it.

You can load the extractor code as a [plugin](https://github.com/yt-dlp/yt-dlp#plugins) into yt-dlp

**Comment by CHJ85** [comment]
@pukkandan  Please add this extractor.
Youtube-dl are being overly careful now ever since they were temperately shut down some time ago.
This is not necessarily a pirate site. And even if it was, adding support for pirate sites is not illegal. By that logic, any search engine site would be shut down because you can use em to find pirated content. And also by that logic, torrent sites would be shut down because you can download pirated stuff from there. But they're being used for legit purposes too.
Adding extractors is like hotlinking.

**Comment by pukkandan** [comment]
I will not be adding this to the main repo. The plugin system exists so that you don't need my permission to add whatever site. You can ask the author to distribute this as a plugin if you want.

**Comment by CHJ85** [comment]
@pukkandan Gotcha. I didn't even consider the plugin option. Now I see that you even mentioned it to him. My bad. 😀
Is there any way you could turn this into a plugin though, please?
I don't know if he's gonna be able to as he just copy/pasted this pull request from youtube-dl's github.

**Comment by pukkandan** [comment]
Converting extractor to plugin is easy.

Just copy paste the extractor into `ytdlp_plugins/extractor` and change the relative imports:

```py
from yt_dlp.extractor import InfoExtractor
from yt_dlp.utils import js_to_json, urljoin
```

Then import it in `ytdlp_plugins/extractor/__init__.py`

See https://github.com/yt-dlp/yt-dlp/blob/master/ytdlp_plugins/extractor for example

**Comment by CHJ85** [comment]
@pukkandan Thank you!

**Comment by CHJ85** [comment]
Btw. I installed yt-dlp using pip3, and the plugins directory is not included in the installation.
You might wanna look into that.

**Pull request closed without merging**